File: DSA/LeetCode/Backtracking/1593. Split a String Into the Max Number of Unique Substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:

        def dfs(i, mSet):

            if i >= len(s):
                return 0
            res = 0 
            for j in range(i, len(s)):

                substr =  s[i:j+1]
                if substr in mSet:
                    continue
                
                mSet.add(substr)
                res = max(res, 1+ dfs(j+1, mSet))
                mSet.remove(substr)
            return res
            
        
        return dfs(0, set())
        

File: DSA/LeetCode/Backtracking/1593. Split a String Into the Max Number of Unique Substrings.js
/**
 * @param {string} s
 * @return {number}
 */
var maxUniqueSplit = function (s) {



    function dfs(i, set) {
        if (i >= s.length)
            return 0

        let res = 0
        for (let j = i; j < s.length; j++) {
            let substr = s.substring(i, j + 1)

            if (set.has(substr))
                continue

            set.add(substr);
            res = Math.max(res, 1 + dfs(j + 1, set))
            set.delete(substr);

        }
        return res
    }

    return dfs(0, new Set());
};